src/kernel/capinv.c

```diff
@@ -1082,6 +1083,76 @@ composite_syscall_slowpath(struct pt_regs *regs)
 		sret_ret(thd, regs, cos_info);
 		return 0;
 	}
+	case CAP_TCAP:
+	{
+		switch (op)
+		{
+		case CAPTBL_OP_TCAP_ACTIVATE:
+		{
+			capid_t tcap_cap     = __userregs_get1(regs) & 0xFFFF;
+			int flags 	     = __userregs_get1(regs) >> 16;
+			capid_t pgtbl_cap    = __userregs_get2(regs);
+			capid_t pgtbl_addr   = __userregs_get3(regs);
+			capid_t compcap      = __userregs_get4(regs);
+
+			struct tcap *tcap_new;
+			unsigned long *pte = NULL;
+
+			ret = cap_kmem_activate(ct, pgtbl_cap, pgtbl_addr, (unsigned long *)&tcap_new, &pte);
+			if (unlikely(ret)) cos_throw(err, ret);
+
+			tcap_split(cap, tcap_new, tcap_cap, ct, compcap, flags);
+			if(ret) {
+				cos_throw(err, ret);
+				unsigned long old = *pte;
+				assert (old & PGTBL_COSKMEM);
+
+				retypetbl_deref((void *)(old & PGTBL_FRAME_MASK));
+				*pte = old & ~PGTBL_COSKMEM;
+			}
+
+			break;
+		}
+		case CAPTBL_OP_TCAP_TRANSFER:
+		{
+			struct tcap *tcapsrc = (struct tcap *)__userregs_getcap(regs);
+			struct tcap *tcapdst = (struct tcap *)__userregs_get1(regs);
+			long long res = __userregs_get2(regs);
+			u32_t prio_higher = __userregs_get3(regs);
+			u32_t prio_lower = __userregs_get4(regs);
+			u16_t prio = 0; /* = higher_prio << 32 || lower_prio; */
+
+			ret = tcap_transfer(tcapdst, tcapsrc, res, prio);
+
+			if (ret) cos_throw(err, -ENOENT);
+
+			break;
+		}
+		case CAPTBL_OP_TCAP_DELEGATE:
+		{
+			struct tcap *tcapsrc = (struct tcap *)__userregs_getcap(regs);
+			struct cap_arcv *arcv = (struct cap_arcv *) __userregs_get1(regs);
```

Comment:
For an example of doing the arcv lookup see the CAP_ARCV handling in the fast-path.